Fix crash when user's locale is unsupported by the spellchecker

tsvetilian-ty opened the pull request:
Closes #14817

## Description
Added an additional check to make sure the user's locale is part of the `availableSpellCheckerLanguages` ([docs](https://www.electronjs.org/docs/latest/api/session#sesavailablespellcheckerlanguages-readonly))

tsvetilian-ty committed:
Fix crash when lang is unsupported by spellcheck

tidy-dev commented:
Beat me to it - I was just starting to take a look at a fix for this. :D

tidy-dev reviewed (changes requested):
Thanks for picking this up!

However, I tested this logic by simply hard coding the user language as `zh-CN` and this still crashes. I believe that is do the this check to return null being a && when it needs to be an ||.

However.. I was starting down the path of making this function unit testable since I can't easily test it on my mac and feel like since there are at least 4 code paths in it, it would be nice to have. Thus, need to make this function accept the things it retrieves from the session as parameters and then added spell-checker-menu-test.ts to the test/main-process directory. Could you update your PR for that? (If not, I can forge on with what I started :)) 

Here are the 4 scenarios I was starting test for off the top of my head:
- spell checker language is english && user language is english -> return null
- spell checker language is english && user language is not english && user language is supported -> return `set spellcheck to system language` menu item
- spell checker language is english && user language is not english && user language is not supported -> return null
- spell checker is not english, return `Set spellcheck to English` menu item.

tsvetilian-ty committed:
Change function signature to make it testable

Change signature and refactor the function to make it more testable

tsvetilian-ty committed:
Add available spellchecker languages as test helper

tsvetilian-ty committed:
Add tests for build-spell-check-menu.ts

tsvetilian-ty commented:
@tidy-dev nice catch!

I changed the function to be more testable and added the four cases you suggested.
I had an issue while running the tests with the return type `MenuItem` `TypeError: electron_1.MenuItem is not a constructor ` I didn't spend too much time on it and ended up using a more generic approach. This way, the function is completely decoupled from any electron dependencies. 

Please share if you think there is a better way of doing it.

tidy-dev reviewed (dismissed):
✨  Thanks for those changes!  I tested it on my windows vm. Set my system language to Spanish and then to the zh-ch that the user reported with. Received the menu items as expected for Spanish and did not receive a menu item for the Chinese (and Chinese defaulted to English per the library electron is using). 

Random thoughts: I am thankful for system iconography. I know enough Spanish to recognize enough words to bounce around.. but Chinese I was lost. :D

sergiou87 commented on app/src/main-process/menu/build-spell-check-menu.ts:
```diff
@@ -104,8 +127,8 @@ function getSpellCheckLanguageMenuItem(
-    click: () => session.setSpellCheckerLanguages([languageCode]),
-  })
+  return {
+    label: label,
```
Nit:
```suggestion
    label,
```

sergiou87 commented on app/src/main-process/menu/build-spell-check-menu.ts:
```diff
@@ -60,35 +60,58 @@ function getSpellCheckMenuItems(
-  session: Electron.Session
-): MenuItem | null {
-  const userLanguageCode = app.getLocale()
+export function getSpellCheckLanguageMenuItemOptions(
```
Since this is now exported, we'd rather have its parameters also documented in the comment above 😳

sergiou87 commented on app/test/unit/main-process/spell-checker-menu-test.ts:
```diff
@@ -0,0 +1,58 @@
+
+describe('spell-checker-menu', () => {
+  it('spell checker language is english && user language is english', () => {
+    const spellcheckerLanguage = ['en-US']
+    const userLanguageCode = 'en-US'
+
+    const spellcheckerMenuItem = getSpellCheckLanguageMenuItemOptions(
+      userLanguageCode,
+      spellcheckerLanguage,
```
Given `userLanguageCode` and `spellcheckerLanguage` are not used anywhere else, you can just get rid of them:
```suggestion
    const spellcheckerLanguage = ['en-US']
    const userLanguageCode = 'en-US'

    const spellcheckerMenuItem = getSpellCheckLanguageMenuItemOptions(
      'en-US',
      ['en-US'],

```
(You might need to run the linter if you accept these changes).

This also applies to the other tests

tsvetilian-ty replied:
@sergiou87 initially I implemented the tests without `userLanguageCode` and `spellcheckerLanguage`. I added them after that to increase readability because the parameters looked too similar even though they are different types.

sergiou87 commented on app/test/unit/main-process/spell-checker-menu-test.ts:
```diff
@@ -0,0 +1,58 @@
+import { getAvailableSpellcheckerLanguages } from '../../helpers/menus/available-spellchecker-languages-helper'
+
+describe('spell-checker-menu', () => {
+  it('spell checker language is english && user language is english', () => {
```
For the test description, we use sentences that describe how the subject under test should behave (where `it` is the subject under test):
```suggestion
  it('returns null menu item options when both spellchecker language and user language are english', () => {

```

sergiou87 commented on app/test/unit/main-process/spell-checker-menu-test.ts:
```diff
@@ -0,0 +1,58 @@
+      getAvailableSpellcheckerLanguages()
+    )
+
+    expect(spellcheckerMenuItem?.label).toBe(
+      'Set spellcheck to system language'
+    )
```
Not extremely concerned about this, but I think it's better if we don't use specific user-facing strings in tests. Maybe we could do something like…
```suggestion
    expect(spellcheckerMenuItem?.label).not.toBeFalsy()
```

That will check the label is not undefined, nor null, nor an empty string.

tidy-dev replied:
I am not sure I agree with this one. I get the idea - but, the label is the result we are checking for. There could be 2 different labels or null. If we want the test to be agnostic, we could put the labels in an exportable constant used by the implementation and the test.

tsvetilian-ty replied:
I refactor the tests and impl to rely on constants as @tidy-dev suggested!

sergiou87 replied:
Sounds good, thank you both! ❤️

sergiou87 reviewed (changes requested):
Thanks for your contribution! Just a few things I spotted 😄

tsvetilian-ty committed:
Add param docs and extracted the labels as const

tsvetilian-ty committed:
Update test descriptions and using the label const

sergiou87 reviewed (approved):
Thanks for those changes 🙌 :shipit:
